api/envoy/service/ext_proc/v3alpha/external_processor.proto

```diff
@@ -0,0 +1,300 @@
+  // received from downstream will be ignored, and any additional messages
+  // from the remote server on either the request or response path will
+  // also be ignored for the duration of this request or response.
+  bool end_of_stream = 4;
+
+  // Clear the route cache for the current request.
+  // This is necessary if the remote server
+  // modified headers that are used to calculate the route.
+  bool clear_route_cache = 5;
+}
+
+// This message must be sent in response to a request_trailers
+// or response_trailers message.
+message TrailersResponse {
+  // Instructions on how to manipulate the trailers
+  HeaderMutation header_mutation = 1;
+}
+
+// TODO this is now the same as HeaderResponse. However, do we think
+// that will always be the case? Keeping it as a separate message
+// for now.
+// [#next-free-field: 6]
+message BodyResponse {
+  // If set, manipulate HTTP headers along with the body. Note that
+  // this is only possible if the headers have not been sent down
+  // the filter chain yet. Practically, then, this field will be ignored
+  // unless the request or response BodySendMode is BUFFERED or
+  // BUFFERED_PARTIAL, which means that Envoy is buffering the body
+  // in memory while waiting for this response.
+  HeaderMutation header_mutation = 1;
+
+  // Request that the last body chunk delivered be modified.
+  BodyMutation body_mutation = 2;
+
+  // If set, add trailers to the message. If and only if this
+  // field is combined with the end_stream flag, then
+  // the trailers will be sent along with the request or
+  // response -- otherwise, this field is ignored.
+  config.core.v3.HeaderMap trailers = 3;
+
+  // If true, then replace the request and response with the contents
+  // of this message. If header_mutation is set and headers have not
+  // already been forwarded down the filter chain, apply it to the
+  // headers. If body_mutation is set and contains a body, then add that
+  // body to the request or response, even if one does not already exist --
+  // otherwise, clear the body. Any additional body and trailers
+  // received from downstream will be ignored, and any additional messages
+  // from the remote server on either the request or response path will
+  // also be ignored for the duration of this request or response.
+  bool end_of_stream = 4;
+
+  // Clear the route cache for the current request.
+  // This is necessary if the remote server
+  // modified headers that are used to calculate the route.
+  bool clear_route_cache = 5;
+}
+
+// This message completely replaces the response. If the response headers
+// have not already been sent downstream, then this will cause the filter to
+// immediately return the data specified here to the downstream
+// system, stop further filter processing, and ignore further messages coming
+// from the remote server. However, if the response headers have already been
+// delivered downstream, then this message will be ignored.
```

Comment:
I think you need way to do a reset though? Do you want to have an explicit reset message, or do you want to use this also as the reset message?

Reply:
Let's make this consistent with "sendLocalReply" on the HTTP filter interface. I'll add the details and change to a "HeaderMutation" structure.

However, FWIW, sendLocalReply doesn't let you set trailers, only the grpc-status. I think I will make that change as well.

Reply:
And by "make this consistent," I mean that sendLocalReply will respond with a new HTTP response if it can and reset the stream if it's too late for that -- I'll adopt that behavior and likely the comment text as well.